src/MacroExpander.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * This file contains the “gullet” where macros are expanded
+ * until only non-macro tokens remain.
+ */
+
+var Lexer = require("./Lexer");
+
+function MacroExpander(input, macros) {
+    this.lexer = new Lexer(input);
+    this.macros = macros;
+    this.stack = []; // contains tokens in REVERSE order
+    this.discardedWhiteSpace = [];
+}
+
+/**
+ * Expand first token while that is expandable,
+ * return first non-expandable token that remains.
```

Comment:
But also incorrect. If you have `\def\foo{\bar\bar}\def\bar{x}` then all the tokens `\foo` generates would be `xx` but it only expands this to `x\bar` then returns the `x`without expanding the second `\bar` at this point.